File: 2024/07/p2.test.lua
local describe = require("lib.test")
local p2 = require(".2024.07.p2")

describe("input", function(it)
  it("example", function()
  local example = [[190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20]]
    local expected = 11387

    local n = p2.solve(example)

    assert(n == expected, string.format("Expected %d, but was %d", expected, n))
  end)

  it("input", function()
    local fp = io.open("./2024/07/assets/input.txt", "r") or error("Input not found")
    local input = fp:read("*a")
    local expected = 20928985450275

    local n = p2.solve(input)

    assert(n == expected, string.format("Expected %d, but was %d", expected, n))
  end)
end)

File: 2024/10/__tests__/p2.test.ts
import assert from "node:assert";
import { readFile } from "node:fs/promises";
import { describe, it } from "node:test";
import { solve } from "../p2.js";
import path from "node:path";

describe("p2", () => {
  it("example 1", () => {
    const example = `.....0.
..4321.
..5..2.
..6543.
..7..4.
..8765.
..9....`;

    assert.strictEqual(solve(example), 3);
  });

  it("example 2", () => {
    const example = `..90..9
...1.98
...2..7
6543456
765.987
876....
987....`;

    assert.strictEqual(solve(example), 13);
  });

  it("example 3", () => {
    const example = `012345
123456
234567
345678
4.6789
56789.`;

    assert.strictEqual(solve(example), 227);
  });

  it("example 4", () => {
    const example = `89010123
78121874
87430965
96549874
45678903
32019012
01329801
10456732`;

    assert.strictEqual(solve(example), 81);
  });

  it("input", async () => {
    const filename = path.join(import.meta.dirname, "../assets/input.txt").replace("/dist/", "/");
    const input = await readFile(filename, "utf8");

    assert.strictEqual(solve(input), 1326);
  });
});
